pkg/controller/persistentvolume/persistentvolume_framework_test.go

```diff
@@ -0,0 +1,1020 @@
+
+var novolumes []*api.PersistentVolume
+var noclaims []*api.PersistentVolumeClaim
+var noevents = []string{}
+var noerrors = []reactorError{}
+
+// volumeReactor is a core.Reactor that simulates etcd and API server. It
+// stores:
+// - Latest version of claims volumes saved by the controller.
+// - Queue of all saves (to simulate "volume/claim updated" events). This queue
+//   contains all intermediate state of an object - e.g. a claim.VolumeName
+//   is updated first and claim.Phase second. This queue will then contain both
+//   updates as separate entries.
+// - Number of changes since the last call to volumeReactor.syncAll().
+// - Optionally, volume and claim event sources. When set, all changed
+//   volumes/claims are sent as Modify event to these sources. These sources can
+//   be linked back to the controller watcher as "volume/claim updated" events.
+// - Optionally, list of error that should be returned by reactor, simulating
+//   etcd / API server failures. These errors are evaluated in order and every
+//   error is returned only once. I.e. when the reactor finds matching
+//   reactorError, it return appropriate error and removes the reactorError from
+//   the list.
+type volumeReactor struct {
+	volumes              map[string]*api.PersistentVolume
+	claims               map[string]*api.PersistentVolumeClaim
+	changedObjects       []interface{}
+	changedSinceLastSync int
+	ctrl                 *PersistentVolumeController
+	volumeSource         *framework.FakeControllerSource
+	claimSource          *framework.FakeControllerSource
+	lock                 sync.Mutex
+	errors               []reactorError
+}
+
+// reactorError is an error that is returned by test reactor (=simulated
+// etcd+/API server) when an action performed by the reactor matches given verb
+// ("get", "update", "create", "delete" or "*"") on given resource
+// ("persistentvolumes", "persistentvolumeclaims" or "*").
+type reactorError struct {
+	verb     string
+	resource string
+	error    error
+}
+
+// React is a callback called by fake kubeClient from the controller.
+// In other words, every claim/volume change performed by the controller ends
+// here.
+// This callback checks versions of the updated objects and refuse those that
+// are too old (simulating real etcd).
+// All updated objects are stored locally to keep track of object versions and
+// to evaluate test results.
+// All updated objects are also inserted into changedObjects queue and
+// optionally sent back to the controller via its watchers.
+func (r *volumeReactor) React(action core.Action) (handled bool, ret runtime.Object, err error) {
+	r.lock.Lock()
+	defer r.lock.Unlock()
+
+	glog.V(4).Infof("reactor got operation %q on %q", action.GetVerb(), action.GetResource())
+
+	// Inject error when requested
+	err = r.injectReactError(action)
+	if err != nil {
+		return true, nil, err
+	}
+
+	// Test did not requst to inject an error, continue simulating API server.
```

Comment:
request (very minor!)
Also, I have a comment or two on your persistentvolume_controller.go file but since is it > 1500 lines github doesn't display it. How can I make these comments?